### PR TITLE
Heyi fused grad accumulation

### PR DESCRIPTION
**Pull request by eliotwang**
hipblaslt implementation in fused_weight_gradient_dense

### CODE REVIEW

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu [hunk 1]
```diff
@@ -10,10 +10,285 @@
...
+    void*     d_workspace,
+    int64_t   max_workspace_size,
+    hipStream_t   stream) {
+
+    hipblaslt_ext::GemmPreference gemmPref;
+    gemmPref.setMaxWorkspaceBytes(max_workspace_size);
+    hipblaslt_ext::Gemm gemm(
+        handle, transa, transb, HIP_R_16BF, HIP_R_16BF, HIP_R_32F, HIP_R_32F, HIPBLAS_COMPUTE_32F);
+
+    hipblaslt_ext::GemmEpilogue
```
Comment (wenchenvincent) [inline csrc/megatron/fused_weight_gradient_dense_cuda.cu:66]: This is using hipblasLtExt API. We haven't used that a lot. We mostly use hipblasLt API. I will need to check with hipblaslt team regarding the difference (in performance implication). And as hipblasLt API is more commonly used, I expect it would be more stable and have fewer bugs.

**Comment by eliotwang** [comment]
update summary:
1. replace hipblasltext API with hipblaslt API
2. create and destroy handle per call
3. add test_weight_grad.py to tests/L0/run_transformer/ following utils test rules

File: csrc/megatron/fused_weight_gradient_dense_16bit_prec_cuda.cu [hunk 2]
```diff
@@ -10,10 +10,231 @@
...
+
+
+//hipblasLtHandle_t g_hipblas_handle = nullptr;
 
+template <typename T>
+void wgrad_gemm_accum_fp16_cuda(T *input, T *d_output,  T *dc_tensor, T *d_weight,int in_dim, int hidden_dim, int out_dim) {
+    //cublasHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+    //if(g_hipblas_handle == nullptr)
+    //   CHECK_HIPBLAS_ERROR(hipblasCreate(&g_hipblas_handle));
+    hipblasLtHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply csrc/megatron/fused_weight_gradient_dense_16bit_prec_cuda.cu]: correct.  cuda handles can be used interchangeably, but hipblas and hipblaslt handles cannot.

File: tests/L0/run_transformer/test_weight_grad.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        #BATCH_SIZE, SEQUENCE_LENGTH, HIDDEN_SIZE, OUTPUT_SIZE
+        (1, 14, 1531, 500),
+        (1, 32, 1536, 256),
+        (1, 64, 3584, 512),
+        (1, 128, 4096, 1024),
+        (8, 32, 1536, 256),
+        (8, 64, 3584, 2048),
+        (8, 128, 4096, 4096),       
+    ]
+    def cosin_similarity(self, vector_a, vector_b):
```
Comment (wenchenvincent) [inline tests/L0/run_transformer/test_weight_grad.py:22]: Is there a specific reason why we use this custom `cosin_similarity`? Can we just use `torch.all_close`?

File: tests/L0/run_transformer/test_weight_grad.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        C_temp = C.float()
+        
+        C_temp.data.mul_(beta)
+        C_temp.data.add_(scaled_AB)
+
+        C.copy_(C_temp).to(dtype=output_type, device="cuda")
+        return C
+
+
+    def RunWeightGradDense(self, ab_type, cd_type):
```
Comment (wenchenvincent) [inline tests/L0/run_transformer/test_weight_grad.py:66]: We will also want to add the combinations of transA and transB in the testing.

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu [hunk 5]
```diff
@@ -10,10 +10,481 @@
...
+        d_weight,   //dd
+        d_workspace,
+        max_workspace_size,
+        stream);
+
+    CHECK_HIPBLASLT_ERROR(hipblasLtDestroy(handle));
+    CHECK_HIP_ERROR(hipStreamDestroy(stream));
+} 
+    
+template void wgrad_gemm_accum_fp32_cuda<at::Half>(at::Half *input, at::Half *d_output, float *dc_tensor, float *d_weight, int in_dim, int hidden_dim, int out_dim);
```
Comment (wenchenvincent) [inline csrc/megatron/fused_weight_gradient_dense_cuda.cu:443]: Do we need to add an extra argument of `dc_tensor` here? We can probably use `d_weight` as the input and output.

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu [hunk 6]
```diff
@@ -10,10 +10,481 @@
...
+    void*   d_workspace,
+    int64_t  max_workspace_size,
+    hipStream_t   stream) {
+    hipblasLtMatrixLayout_t matA, matB, matC, matD;
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutCreate(&matA, HIP_R_32F, m, k, m));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutCreate(&matB, HIP_R_32F, n, k, n));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutCreate(&matC, HIP_R_32F, m, n, m));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutCreate(&matD, HIP_R_32F, m, n, m));
+
+    if(batch_count > 1)
```
Comment (wenchenvincent) [inline csrc/megatron/fused_weight_gradient_dense_cuda.cu:309]: As we already have folded the batch dimension at higher level calls, we don't need to deal with `batch_count>1` in the `gemmex_wrapper` and only need to deal with `batch_count=1` for simplicity.

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu [hunk 7]
```diff
@@ -10,10 +10,481 @@
...
+        CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutSetAttribute(
+            matD, HIPBLASLT_MATRIX_LAYOUT_BATCH_COUNT, &batch_count, sizeof(batch_count)));
+        CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutSetAttribute(
+            matD, HIPBLASLT_MATRIX_LAYOUT_STRIDED_BATCH_OFFSET, &stride_d, sizeof(stride_d)));
+    }
+
+    hipblasLtMatmulDesc_t matmul;
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescCreate(&matmul, HIPBLAS_COMPUTE_32F, HIP_R_32F));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_TRANSA, &transa, sizeof(int32_t)));
```
Comment (wenchenvincent) [inline csrc/megatron/fused_weight_gradient_dense_cuda.cu:336]: Not sure if `transa` is `int32_t` or not. But I guess we can just use `sizeof(transa)`.

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu [hunk 8]
```diff
@@ -10,10 +10,481 @@
...
+    CHECK_HIPBLASLT_ERROR(
+        hipblasLtMatmulPreferenceSetAttribute(pref,
+                                              HIPBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES,
+                                              &max_workspace_size,
+                                              sizeof(max_workspace_size)));
+
+    const int                        request_solutions = 1;
+    hipblasLtMatmulHeuristicResult_t heuristicResult[request_solutions];
+    int                              returnedAlgoCount = 0;
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulAlgoGetHeuristic(handle,
```
Comment (wenchenvincent) [inline csrc/megatron/fused_weight_gradient_dense_cuda.cu:356]: @jeffdaily Do we do any autotuning on the hipblasLt path? I am wondering if we need to do autotuning here.

File: tests/L0/run_transformer/test_weight_grad.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        for BATCH_SIZE, SEQUENCE_LENGTH, HIDDEN_SIZE, OUTPUT_SIZE in self.configs:
+            print(f"BATCH_SIZE = {BATCH_SIZE}, SEQUENCE_LENGTH = {SEQUENCE_LENGTH}, HIDDEN_SIZE = {HIDDEN_SIZE}, OUTPUT_SIZE = {OUTPUT_SIZE}")
+            input_shape = (SEQUENCE_LENGTH, BATCH_SIZE, HIDDEN_SIZE) # k*m
+            output_shape = (SEQUENCE_LENGTH, BATCH_SIZE, OUTPUT_SIZE) # k*n
+            weight_shape = (OUTPUT_SIZE, HIDDEN_SIZE)
+            
+            torch.manual_seed(self.SEED)
+
+            input = torch.randn(input_shape, requires_grad=False, device="cuda") # A
+            output = torch.randn(output_shape, requires_grad=False, device="cuda") #B
```
Comment (wenchenvincent) [inline tests/L0/run_transformer/test_weight_grad.py:76]: Nit: here output is actually gradient of output and weight is actually gradient of weight. So if the name can reflect that, it will avoid confusion.

File: tests/L0/run_transformer/test_weight_grad.py (hunk 4 above)
Comment (wenchenvincent) [reply tests/L0/run_transformer/test_weight_grad.py:66]: I take this back. For the function at this level, the transposition is already fixed. Only if we test the gemmex_wrapper, then we need to test combinations of transa and transb.

File: tests/L0/run_transformer/test_weight_grad.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
+import torch
+import fused_weight_gradient_mlp_cuda
+import torch.nn as nn
+import sys
+import math, pdb
+from torch.testing._internal import common_utils
+
+torch.backends.cuda.matmul.allow_tf32 = False
```
Comment (wenchenvincent) [inline tests/L0/run_transformer/test_weight_grad.py:8]: @jeffdaily Do we have this default to be False on rocm?

**Comment by eliotwang** [comment]
update summary:

1. use getCurrentCUDABlasLtHandle to get hipblaslt handle;
2. Update test files to accommodate to the unittest infrastructure and style used by apex; https://github.com/eliotwang/apex/blob/heyi_fused_grad_accumulation/tests/L0/run_transformer/test_weight_grad.py
run test with: python tests/L0/run_test.py --include run_transformer
3. Replace custom cosin_similarity with torch.all_close, set tolerance threshold for each test case
4. Set TENSILE_DB=0x8000 to tell different func call between hipblas and hipblaslt;
hipblas:
Cijk_Ailk_Bjlk_BBS_BH_MT128x64x16_MI16x16x16x1_SN_1LDSB0_APM1_ABV0_ACED0_AF0EM1_AF1EM1_AMAS3_ASE_ASGT_ASLT_ASM_ASAE01_ASCE01_ASEM1_AAC0_BL1_BS1_CLR1_DTLA0_DTLB0_DTVA0_DTVB0_DVO0_ETSP_EPS1_ELFLR0_EMLL0_FSSC10_FL0_GLVWA4_GLVWB2_GRCGA1_GRCGB1_GRPM1_GRVWn1_GSU1_GSUASB_GLS0_ISA942_IU1_K1_KLA_LBSPPA0_LBSPPB512_LPA0_LPB16_LDL1_LRVW4_LWPMn1_LDW0_FMA_MIAV0_MDA2_MO40_MMFSC_MKFGSU256_NTA0_NTB0_NTC0_NTD0_NEPBS0_NLCA2_NLCB2_ONLL1_OPLV0_PK0_PAP0_PGR2_PLR1_PKA0_SIA3_SLW1_SS1_SU32_SUM0_SUS256_SCIUI1_SPO0_SRVW0_SSO0_SVW4_SNLL0_TSGRA0_TSGRB0_TT8_16_TLDS0_UMLDSA0_UMLDSB0_USFGROn1_VAW1_VSn1_VW4_VWB1_VFLRP1_WSGRA0_WSGRB0_WS64_WG16_16_1_WGMn16
hipblaslt:
Cijk_Ailk_Bjlk_BBS_BH_Bias_AS_SAV_UserArgs_MT64x32x32_MI16x16x1_SN_LDSB0_AFC1_AFEM1_AFEM1_ASEM1_CLR1_CADS0_EPS0_GRVWA8_GRVWB4_GSUAMB_ISA942_IU1_K1_LBSPPA512_LBSPPB256_LBSPPM0_LPA32_LPB16_LPM0_LRVW4_LWPMn1_MIAV0_MIWT2_1_MO40_NTn1_NTA0_NTB0_NTC0_NTD0_NTM0_NEPBS16_NLCA1_NLCB1_ONLL1_PGR2_PLR1_PKA1_SIA3_SS1_SPO0_SRVW0_SSO0_SVW2_TLDS0_USFGROn1_VSn1_VWA2_VWB1_WSGRA0_WSGRB0_WS64_WG32_8_1
5. Remove dc_tensor, use d_weight as both the input and output;
6. Remove batch_count>1 case;
7. Replace sizeof(int32_t) sizeof(transa);
8. Rename vars as gradient of output and gradient of weight to avoid confusion;

File: tests/L0/run_transformer/test_weight_grad.py [hunk 11]
```diff
@@ -55,81 +40,72 @@ def backward_matrix_operation(self, alpha, beta, A, B, C):
...
 
-            #compare
-            similarity = self.cosin_similarity(weight, ref_weight)
-            print("similarity :", similarity)
-            if similarity > 0.999:
-                print("========PASS======")
-            else:
-                print("========FAIL======")
-               
+            grad_weight = grad_weight.view(-1)
```
Comment (wenchenvincent) [inline tests/L0/run_transformer/test_weight_grad.py:93]: We don't need to reshape the tensor now as `torch.allclose` can take 2d tensors.

**Comment by wenchenvincent** [comment]
@pruthvistony @jithunnair-amd Do we have any CI setup for Apex?

**Comment by pragupta** [comment]
> @pruthvistony @jithunnair-amd Do we have any CI setup for Apex?

@wenchenvincent -- there's no CI for apex.

**Comment by pragupta** [comment]
Regarding the tolerance for UTs, I looked at matmul UTs in PyTorch and it seems there are some `toleranceOverride` decorators which define the tolerance level for various operators and dtypes. Pleasea see lines like these in this file: https://github.com/pytorch/pytorch/blob/68272ab5967f448ed6d2986039a0ef0ddf0e1b37/test/test_matmul_cuda.py#L119

**Comment by amd-sriram** [comment]
already merged with different PR https://github.com/ROCm/apex/pull/183
